apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/PriorityListTests.cs

```diff
@@ -0,0 +1,235 @@
+            Assert.False(itemToAdd.HasEventHooks);
+        }
+
+        [Fact]
+        public void MultipleAddRemovesStaySorted()
+        {
+            PriorityList<TestItem> priorityList = new PriorityList<TestItem>();
+            TestItem item1, item2, item3, item4, item5;
+
+            priorityList.Add(item1 = new TestItem(1));
+            priorityList.Add(item2 = new TestItem(5));
+            priorityList.Add(item3 = new TestItem(3));
+            priorityList.Add(item4 = new TestItem(5));
+            priorityList.Add(item5 = new TestItem(6));
+
+            AssertExist(priorityList, item1, item2, item3, item4, item5);
+            AssertSorted(priorityList);
+
+            priorityList.Remove(item3);
+            AssertNotExist(priorityList, item3);
+            AssertSorted(priorityList);
+
+            priorityList.Remove(item4);
+            AssertNotExist(priorityList, item4);
+            AssertExist(priorityList, item2);
+            AssertSorted(priorityList);
+
+            priorityList.Add(item4);
+            AssertExist(priorityList, item4);
+            AssertSorted(priorityList);
+            Assert.Equal(4, priorityList.Count);
+        }
+
+        [Fact]
+        public void GetTop()
+        {
+            PriorityList<TestItem> priorityList = new PriorityList<TestItem>();
+
+            priorityList.Add(new TestItem(3));
+            priorityList.Add(new TestItem(11));
+            priorityList.Add(new TestItem(2));
+            priorityList.Add(new TestItem(0));
+            priorityList.Add(new TestItem(6));
+
+            AssertSorted(priorityList);
+
+            Assert.Equal(0, priorityList.GetTop().Priority);
+        }
+
+        [Fact]
+        public void RemoveLast()
+        {
+            PriorityList<TestItem> priorityList = new PriorityList<TestItem>();
+            TestItem itemToRemove;
+
+            priorityList.Add(new TestItem(3));
+            priorityList.Add(itemToRemove = new TestItem(11));
+            priorityList.Add(new TestItem(2));
+            priorityList.Add(new TestItem(0));
+            priorityList.Add(new TestItem(6));
+
+            priorityList.RemoveLast();
+            AssertSorted(priorityList);
+            AssertNotExist(priorityList, itemToRemove);
+        }
+
+        [Fact]
+        public void PriorityChangeResortsList()
+        {
+            PriorityList<TestItem> priorityList = new PriorityList<TestItem>();
+            TestItem changingItem;
+
+            priorityList.Add(new TestItem(3));
+            priorityList.Add(new TestItem(11));
+            priorityList.Add(changingItem = new TestItem(2));
+            priorityList.Add(new TestItem(0));
+            priorityList.Add(new TestItem(6));
+            AssertSorted(priorityList);
+            changingItem.Priority = 20;
+            AssertNotSorted(priorityList);
+            changingItem.FirePriorityChanged();
+            AssertSorted(priorityList);
+        }
```

Comment:
Possible additional test: it may be helpful to do another version of this which changes the priority to something equal to one of the others. Then remove one of the conflicting items after the `FirePriorityChanged` to make sure the correct one is removed.